### ISSUE TITLE
how to maintain a git remote for a sd-scripts PR for passing args to WAND

### ISSUE THREAD

**Issue by rafstahelin** [bmaltais/kohya_ss#2482]
I'd like to integrate @gshawn3's sd-scripts PR `hyperparams` which passes args to wandb. 

I have:

`git fetch origin pull/1285/head:hyperparams
git checkout hyperparams`

But in the kohya_ss readme for maintaining a remote git 
`git remote add sd-scripts https://github.com/kohya-ss/sd-scripts.git` 

[https://github.com/bmaltais/kohya_ss/blob/master/examples/pull kohya_ss sd-scripts updates in.md#updating-a-local-branch-with-the-latest-sd-scripts-changes](https://github.com/bmaltais/kohya_ss/blob/master/examples/pull%20kohya_ss%20sd-scripts%20updates%20in.md#updating-a-local-branch-with-the-latest-sd-scripts-changes)

So I run:
```
git remote add sd-scripts https://github.com/kohya-ss/sd-scripts.git
git fetch origin pull/1285/head:hyperparams
```
I get:
`E:\kohya_ss>git fetch origin pull/1285/head:hyperparams
fatal: couldn't find remote ref pull/1285/head`

What is the correct way to fetch and checkout?

**Comment by rafstahelin** [comment]
Currently:
```
E:\kohya_ss>git remote -v
origin  https://github.com/bmaltais/kohya_ss.git (fetch)
origin  https://github.com/bmaltais/kohya_ss.git (push)
sd-scripts      https://github.com/kohya-ss/sd-scripts.git (fetch)
sd-scripts      https://github.com/kohya-ss/sd-scripts.git (push)
```

With 
`git fetch sd-scripts pull/1285/head:hyperparams`

Unsure how checkout and merge given the remote is a PR.

**Comment by ccharest93** [comment]
I heavily recommend the "pro Git" book by Scott Chacon it's available online for free and demystifies a lot surrounding git branching, merging etc

**Comment by rafstahelin** [comment]
Will give this a read as we speak thanks

**Comment by rafstahelin** [comment]
Maybe I should git fetch in the ss-scripts path within kohya_scripts? Maybe it's as plain as that

**Comment by ccharest93** [comment]
There are two repositories you are trying to change. 
You want to change sd_scripts checkout bfb352b to merge the hyperparameter pull request.
You want to change kohya_ss to have its sd_script remote be the branch you created in the previous step.

**Comment by ccharest93** [comment]
Key here is understanding most PR are based on forks, so the hyperparamater PR is not located within the official sd_script tree, it is located in a branch i created from a fork of sd_script (these are two different git trees). You could set your remote to be that branch on my git tree, but by doing that you make yourself dependent on me doing the rebasing when it is needed. It is much better for you to create your own fork of sd_script, checkout the bfb352b:commit, create a branch there, merge the hyperparameter PR within that branch.

THEN you can come to the ss_kohya, and modify it to use that branch as the remote for sd_script instead of the bfb352b:commit.

**Comment by rafstahelin** [comment]
Think I understand. I've managed to push the merge to my forked branch. Now I need to add it as remote in place of the commit.
Testing

**Comment by rafstahelin** [comment]
@ccharest93  

here is my PR. It's my first one.
[(https://github.com/kohya-ss/sd-scripts/pull/1327)]
Hope this is right. I can see your PR is correctly reflected 

I am now proceeding to step 2, set it as remote on kohya

**Comment by ccharest93** [comment]
oups i dont think you want to open a PR with your branch on the sd_script repo.

**Comment by rafstahelin** [comment]
> oups i dont think you want to open a PR with your branch on the sd_script repo.

Total noobie. I'll remove it
And I'll set the remote to my fork
No need of a PR indeed

**Comment by ccharest93** [comment]
A pull request is meant to be when you want the changes you made into your own copy of the repository(the fork) to be also made into the main repository(official sd_script). This is not the case here, what you did is good up until the point of opening a PR request( which you should close, since my PR is already opened there and reflects those changes).

**Comment by gshawn3** [comment]
Thanks @ccharest93, sorry raf I couldn't chime in earlier as I got caught up with work today!

Hopefully kohya merges the hyperparams PR soon and this little dance will no longer be needed 🙂

**Comment by rafstahelin** [comment]
No worries, in the end I managed to setup the forked hypatams into my locale. Learned a lot from it. Smart system

**Comment by rafstahelin** [comment]
> A pull request is meant to be when you want the changes you made into your own copy of the repository(the fork) to be also made into the main repository(official sd_script). This is not the case here, what you did is good up until the point of opening a PR request( which you should close, since my PR is already opened there and reflects those changes).

@ccharest93 all sorted. Learned the abc's. Thanks for your help

**Comment by rafstahelin** [comment]
Hey @gshawn3 i noticed kohya merged the PR https://github.com/kohya-ss/sd-scripts/pull/1285#event-12857774925
Tried training in Kohya_ss with sd-scripts set to dev, but didnt see parameters passing. Does it work for you?

**Comment by gshawn3** [comment]
Hi, yes it is working correctly for me:

![image](https://github.com/bmaltais/kohya_ss/assets/133769806/582b9061-eab9-4468-904c-f23fb3332232)

Note that I'm running the training sd-script directly from the command line, where I added the new `--log_config` argument to enable this  functionality. (In bmaltais' GUI, if you click on the "Print training command" button at the bottom of the screen, you'll see the full command line with all the arguments that you can subsequently modify and use in your terminal.)

**Comment by rafstahelin** [comment]
Just to be sure, you are using the PR Hyperparams or the latest Dev branch? 

Ok, I understand, running kohya_ss with ss-scripts in set dev branch remote, and running the training in cli with the --log config arg 

Will test thx

**Comment by gshawn3** [comment]
Yeah just to confirm, I'm using the latest dev branch. Good luck!

**Comment by rafstahelin** [comment]
Hi George, pardon me for the simple question.

- I switched to sd-scripts remote on dev branch
- I ran with --log_config in CLI with toml:

`(venv) E:\kohya_ss>E:\kohya_ss\venv\Scripts\accelerate.EXE launch --dynamo_backend no --dynamo_mode default --mixed_precision bf16 --num_processes 1 --num_machines 1 --num_cpu_threads_per_process 4 E:/kohya_ss/sd-scripts/sdxl_train.py --config_file "E:\studio Dropbox\studio\ai\data\styles\style_lighting_graphicwhite\3training\training_gr4ph1cw\models\config_finetune-20240520-120554.toml" --log_config`

And tried with log_config = true in toml alternatively:
![image](https://github.com/bmaltais/kohya_ss/assets/64715158/748d6f7b-675f-4c8d-89b3-935692c075c3)

but neither seems to work

If I copy the Print commands from console, it has the problematic line escape format
![image](https://github.com/bmaltais/kohya_ss/assets/64715158/a091cb33-dca7-4e7c-ba77-757b1b5397a7)

Did you use this last to run?

**Comment by rafstahelin** [comment]
> Hi, yes it is working correctly for me:
> 
> ![image](https://private-user-images.githubusercontent.com/133769806/331988070-582b9061-eab9-4468-904c-f23fb3332232.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTYyMDE1MjMsIm5iZiI6MTcxNjIwMTIyMywicGF0aCI6Ii8xMzM3Njk4MDYvMzMxOTg4MDcwLTU4MmI5MDYxLWVhYjktNDQ2OC05MDRjLWYyM2ZiMzMzMjIzMi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNTIwJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDUyMFQxMDMzNDNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT04OWE2MDU0NWFiYjdhMDJlMzYyYTFhMTJlZGEzZmY0Njk3NTc5ZDc0YzE3MThiZDk4NmQwNWQ0MGE2YjVjODAxJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.YOUGdts9SxPLP0pn6mzN7pu1JNGtvcqJjhB5XEloa2w)
> 
> Note that I'm running the training sd-script directly from the command line, where I added the new `--log_config` argument to enable this functionality. (In bmaltais' GUI, if you click on the "Print training command" button at the bottom of the screen, you'll see the full command line with all the arguments that you can subsequently modify and use in your terminal.)

@gshawn3 
Print Command actually gives you the contents of the toml file, so this is not usable in cli. 

i've tried adding log_config = true in toml, but that doesnt seem to work. 

And just to confirm, You are on kohya_ss main sd-scripts dev

**Comment by rafstahelin** [comment]
@gshawn3 Confirming it works fine with --log_config in cli with toml file. All sorted. Thanks again

**Comment by gshawn3** [comment]
You're welcome, glad to hear you were able to get it sorted out!

**Comment by rafstahelin** [comment]
@gshawn3 hey. I noticed "Additional Arguments" is not passing to wandb. It just shows '[]' whether there is arguments or  not. Know if there is a bug or how to make sure these are sent? I am using additional arguments a lot in gui, but no way to know what was used outside of toml

**Comment by ccharest93** [comment]
accelerate tracker does not support list arguments,
| #Accelerate values need to have type `bool`,`str`, `float`, `int`, or `None`.
as per it's documentation.

**Comment by ccharest93** [comment]
The implement value right now is to convert all objects to string before passing them to the tracker. This is done using format string f"{obj}" which calls the obj.__str__() method. Im assuming that the Additional Arguments array elements don't have a __str__() method declared so the array shows up empty.

**Comment by rafstahelin** [comment]
@ccharest93 thats interesting. But "Optimiser Args" does pass. It's a pity, so arbitrary. Who could change this? Is this a kohya-ss thing?

**Comment by ccharest93** [comment]
You would have to find where the "Aditional arguments" are defined and either change the type of the elements to one that has a \_\_str__() method or defined it yourself on their class. Setting breakpoints with the python debugger to narrow where the argument you are investigating is defined is how i would go about this.

**Comment by rafstahelin** [comment]
@ccharest93 i am also noticing that the annealing I am passing (`--lr_scheduler_type CosineAnnealingLR --lr_scheduler_args T_max=510`) is not actually being passed to the training itself. I wonder if there is a problem. Not just an issue with passing to wandb. I am testing the same training with this additional parameter on main branch to see if this is an issue with dev branch of sd-scripts

**Comment by rafstahelin** [comment]
@ccharest93 just an update. I am running scheduler args for annealing through cli which works

**Comment by bmaltais** [comment]
I have pushed the `--log_config` feature to the dev branch and merged sd-scripts latest dev branch into mine... give it a try.

I also added the other new features that were implemented in kohya's sd-scripts branch in the GUI.

**Comment by rafstahelin** [comment]
> I have pushed the `--log_config` feature to the dev branch and merged sd-scripts latest dev branch into mine... give it a try.
> 
> 
> 
> I also added the other new features that were implemented in kohya's sd-scripts branch in the GUI.

Ok great will try that tomorrow thanks!

**Comment by rafstahelin** [comment]
Hi @bmaltais 
Just noticed --log_config is not in the lora module in dev branch
Could you please add?
thanks

**Comment by rafstahelin** [comment]
@bmaltais how do i set log_config to true in config.toml?

i used "wandb_log_config = true" but the checkbox wasnt x'ed on restart. 

thanks
